tests/test_code_hasher.py

```diff
@@ -0,0 +1,126 @@
+import pytest
+import threading
+
+from bionic.code_hasher import CodeHasher, TypePrefix
+
+
+def test_code_hasher():
+    def barray(value):
+        return bytearray(value, "utf8")
+
+    circular_dict_1_a = {"k11": "v1"}
+    circular_dict_2_a = {"k21": "v2"}
+    circular_dict_3_a = {"k31": "v3"}
+    circular_dict_1_a["k12"] = circular_dict_2_a
+    circular_dict_2_a["k22"] = circular_dict_3_a
+    circular_dict_3_a["k32"] = circular_dict_1_a
+
+    circular_dict_1_b = {"k11": "v1"}
+    circular_dict_2_b = {"k21": "v2"}
+    circular_dict_3_b = {"k31": "v3"}
+    circular_dict_1_b["k12"] = circular_dict_2_b
+    circular_dict_2_b["k22"] = circular_dict_3_b
+    circular_dict_3_b["k32"] = circular_dict_2_b
+
+    def f1():
+        v = 10
+        return v
+
+    def f2():
+        v = 20
+        return v
+
+    def f3():
+        v = "10"
+        return v
+
+    def f4():
+        return "10"
+
+    def inc(x):
+        return x + 1
+
+    def dec(x):
+        return x - 1
+
+    values = [
+        b"",
+        b"123",
+        b"None",
+        barray("bytearray"),
+        barray("anotherbytearray"),
+        None,
+        "",
+        "None",
+        "String1",
+        "String2",
+        "0",
+        "1",
+        "123",
+        "1.23",
+        0,
+        1,
+        123,
+        23,
+        1.23,
+        23.0,
+        float("inf"),
+        float("-inf"),
+        float("nan"),
+        True,
+        False,
+        [1, 2, 3],
+        [1, 2, "3"],
+        [1, 2],
+        (1, 2, 3),
+        (1, 2, "3"),
+        (1, 2),
+        {0: "v1", 1: None, "2": ["value1", "value2"]},
+        {0: "v1", 1: None, "2": ["value1", "value2"], None: "none_val"},
+        {
+            0: "v1",
+            1: {10: "v2", 20: {100: [200, 300]}},
+            "2": ["value1", "value2"],
+            None: "none_val",
+        },
+        circular_dict_1_a,
+        circular_dict_1_b,
+        f1,
+        f2,
+        f3,
+        f4,
+        inc,
+        dec,
+    ]
+
+    idx_by_hash_value = {}
+    for idx, val in enumerate(values):
+        hash_value = CodeHasher.hash(val)
+        # Hashing again should return the same hash value.
+        assert CodeHasher.hash(val) == hash_value
+        assert (
+            hash_value not in idx_by_hash_value
+        ), f"{values[idx]} and {values[idx_by_hash_value[hash_value]]} have the same hash"
+        idx_by_hash_value[hash_value] = idx
+
+
+def test_complex_type_warning():
+    val = threading.Lock()
+    with pytest.warns(
+        UserWarning, match="Found a constant",
+    ):
+        assert CodeHasher.hash(val) == CodeHasher.hash(TypePrefix.DEFAULT)
```

Comment:
Is the fact that we use `TypePrefix.DEFAULT` an important property that we should be testing?